### PR TITLE
FINERACT-2081: Validation error in mandatory loan account parameters

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

In CreateLoan request, missing mandatory fields which were being returned as 400 in the past are currently being returned as 403 ? Is this an expected change? Please confirm. The following fields are being returned with 403 code when missing

- clientId
- productId
- principal
- transactionProcessingStrategyCode
- expectedDisbursementDate
- submittedOnDate

[FINERACT-2081](https://issues.apache.org/jira/browse/FINERACT-2081)

Screenshot 2024-08-30 at 2 17 35 p m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 1]
```diff
@@ -2127,4 +2139,25 @@ public static void validateOrThrow(String resource, Consumer<DataValidatorBuilde
...
+
+    private void setOfficeId(Long officeId, Client client, Group group) {
+        if (client != null) {
+            officeId = client.getOffice().getId();
+        }
+        if (group != null) {
+            officeId = group.getOffice().getId();
+        }
+
+        if (officeId == null) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: I dont really understand why should we throw `clientId` is mandatory if the `officeId` is null. Please review this logic and rewrite to be more straightforward... also setOfficeId method which does not set anything is weird.

Also please be aware if you override the officeId inside this method it has absolutely no any effect outside of this method...

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DataValidatorBuilder.java [hunk 2]
```diff
@@ -1056,4 +1056,12 @@ public void throwValidationErrors() throws PlatformApiDataValidationException {
             throw new PlatformApiDataValidationException(dataValidationErrors);
         }
     }
+
+    public static ApiParameterError buildValidationParameterApiError(final String resource, final String parameterName,
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DataValidatorBuilder.java:1060]: I dont think we need this. You can just simply force return and the `baseDataValidator` would be processed till that point...

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!
Also please add test cases to cover the "correct" behaviour

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Done! update it

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DataValidatorBuilder.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DataValidatorBuilder.java:1060]: I've moved almost all the cases, just one to use this new method

**Comment by jdailey** [comment]
I just note that this is being done on an open ended ticket for "small bug fixes" (2081).... is this a new functionality or a bug fix and if a bug fix the idea should be to make sure the approach is clear, maintainable by the next person coming along.  If the logic and approach is not clear then that defeats the purpose of the changes.   The definition of "DONE" needs to be better than we found it.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 3]
```diff
@@ -1750,9 +1774,6 @@ private void officeSpecificLoanProductValidation(final Long productId, final Lon
     private void validateTransactionProcessingStrategy(final String transactionProcessingStrategy, final LoanProduct loanProduct,
             final DataValidatorBuilder baseDataValidator) {
 
-        baseDataValidator.reset().parameter(LoanApiConstants.transactionProcessingStrategyCodeParameterName)
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java:1753]: why to remove this validation?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 4]
```diff
@@ -2127,4 +2148,22 @@ public static void validateOrThrow(String resource, Consumer<DataValidatorBuilde
                     dataValidationErrors);
         }
     }
+
+    private Long setOfficeId(Client client, Group group) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: It still not sets anything. Please rename it! Example: `fetchOfficeId` or something would be more descriptive...

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!
Also please add test cases to cover the "correct" behaviour

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 5]
```diff
@@ -466,6 +469,9 @@ private void validateForCreate(final JsonElement element) {
                     .extractLocalDateNamed(LoanApiConstants.expectedDisbursementDateParameterName, element);
             baseDataValidator.reset().parameter(LoanApiConstants.expectedDisbursementDateParameterName).value(expectedDisbursementDate)
                     .notNull();
+            if (expectedDisbursementDate == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Alberto, can you pls explain this logic? One line above we're requiring the expectedDisbursementDate to be not null and here we are checking whether it's null and completely ignoring what's coming after that.

Isn't this a bypass to the validation logic (which would follow these lines)?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 6]
```diff
@@ -511,8 +517,10 @@ private void validateForCreate(final JsonElement element) {
 
             final LocalDate submittedOnDate = this.fromApiJsonHelper.extractLocalDateNamed(LoanApiConstants.submittedOnDateParameterName,
                     element);
-
             baseDataValidator.reset().parameter(LoanApiConstants.submittedOnDateParameterName).value(submittedOnDate).notNull();
+            if (submittedOnDate == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Same as above.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 7]
```diff
@@ -523,7 +531,11 @@ private void validateForCreate(final JsonElement element) {
 
             final String transactionProcessingStrategy = this.fromApiJsonHelper
                     .extractStringNamed(LoanApiConstants.transactionProcessingStrategyCodeParameterName, element);
-
+            baseDataValidator.reset().parameter(LoanApiConstants.transactionProcessingStrategyCodeParameterName)
+                    .value(transactionProcessingStrategy).notNull();
+            if (transactionProcessingStrategy == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Same as above.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 4 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Very much agreed. I wouldn't even say "fetch" because it suggests fetching from an external source (like the DB).

I'd rather call it something like "resolveOfficeId". Also, probably this wouldn't be the place to put such a method because in my opinion this is a very generic way to get the office ID for a specific client.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 8]
```diff
@@ -1410,7 +1431,10 @@ public void validateForModify(final JsonCommand command, final Loan loan) {
 
             // validate if disbursement date is a holiday or a non-working day
             validateDisbursementDateIsOnNonWorkingDay(expectedDisbursementDate);
-            Long officeId = client != null ? client.getOffice().getId() : group.getOffice().getId();
+            final Long officeId = setOfficeId(client, group);
+            if (officeId == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Same as above.

**Review by galovics — CHANGES_REQUESTED** [review]
In general I'm missing some tests to cover these changes.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java:1753]: If we keep that validation we get a different error when the the `transactionProcessingStrategy` is null

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Done! method name and logic updated

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: The reason to add this return  there, is because previously we were generating a HTTP 400 Bad Request when some required field like expectedDisbursementDate or principal or transactionProcessingStrategy, etc are missed. 

Now with the new way, we have those schema validations and other business logic mixed, then the business validations are returning first than schema validations.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 6 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: The reason to add this return  there, is because previously we were generating a HTTP 400 Bad Request when some required field like expectedDisbursementDate or principal or transactionProcessingStrategy, etc are missed. 

Now with the new way, we have those schema validations and other business logic mixed, then the business validations are returning first than schema validations.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 7 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: The reason to add this return  there, is because previously we were generating a HTTP 400 Bad Request when some required field like expectedDisbursementDate or principal or transactionProcessingStrategy, etc are missed. 

Now with the new way, we have those schema validations and other business logic mixed, then the business validations are returning first than schema validations.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 8 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: The reason to add this return  there, is because previously we were generating a HTTP 400 Bad Request when some required field like expectedDisbursementDate or principal or transactionProcessingStrategy, etc are missed. 

Now with the new way, we have those schema validations and other business logic mixed, then the business validations are returning first than schema validations.

**Review by alberto-art3ch — COMMENTED** [review]
Please review

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 5 above)
Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Wait @alberto-art3ch. Are we running business validations before schema validations? Shouldn't it be the other way around? 

If we see schema-wise that something is missing, has negative value, etc, how can a business validator even run?

Comment (galovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: One more thing, can you point me to this "new way" schema validations and business validations?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java [hunk 9]
```diff
@@ -172,6 +172,83 @@ public void uc2() {
...
+                    .loanTermFrequencyType(2)//
+                    .transactionProcessingStrategyCode(LoanProductTestBuilder.ADVANCED_PAYMENT_ALLOCATION_STRATEGY)//
+                    .repaymentEvery(1)//
+                    .repaymentFrequencyType(2)//
+            ;//
+
+            callFailedRuntimeException = Assertions.assertThrows(CallFailedRuntimeException.class,
+                    () -> loanTransactionHelper.applyLoan(applicationRequest03));
+
+            Assertions.assertTrue(callFailedRuntimeException.getMessage().contains("The parameter `submittedOnDate` is mandatory."));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java:245]: Would you mind checking the HTTP code?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my review!

**Comment by alberto-art3ch** [comment]
> Please see my review!

Done! Comments reviewed and attended

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java (hunk 5 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: Let me join to the conversation. Before the changes there were validations all over the place, "randomly" through the lifecycle of an action: 
- Some validations for repayment when the service was called first steps
- Some validations in the loan
- Some validations later in the service call
etc.

It was started to rearrange for a couple actions (like loan creation, approval, disbursal) but seems one of the validations type was just collecting "validations issues" in a list and at the end it was evaluated and all errors were thrown eith Http 400.

Now all the validations are in one place seems they got mixed and there were some validations that were throwing http 403 (domain rule exceptions) and this kind of error is immediately thrown, not collected.

To avoid breaking changes our intention was to not change any of the "legacy" validations and which exception is to be thrown, but unfortunately there was some regression (see Alberto's reply), so now his intention is to fix it.

On long run it would be best to have a conversarion on the DEV mail list regarding exception handling but for now, we are aiming to fix the regression.

i hope it helps to gain some extra insight.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java [hunk 10]
```diff
@@ -172,6 +172,80 @@ public void uc2() {
...
+                    operationDate, 100.0, 5).numberOfRepayments(6)//
+                    .submittedOnDate(null) //
+                    .loanTermFrequency(6)//
+                    .loanTermFrequencyType(2)//
+                    .transactionProcessingStrategyCode(LoanProductTestBuilder.ADVANCED_PAYMENT_ALLOCATION_STRATEGY)//
+                    .repaymentEvery(1)//
+                    .repaymentFrequencyType(2)//
+            ;//
+            loanTransactionHelper.applyLoanWithError(applicationRequest03, 400);
+            callFailedRuntimeException = Assertions.assertThrows(CallFailedRuntimeException.class,
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java:227]: Please dont do this.
Use the following to check: `assertEquals(400, callFailedRuntimeException.getResponse().code());`

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 11]
```diff
@@ -3158,7 +3158,7 @@ public void uc121() {
                     .repaymentEvery(15).repaymentFrequencyType(0).numberOfRepayments(3).loanTermFrequency(45).loanTermFrequencyType(0)
                     .principal(BigDecimal.valueOf(1000.0)).maxOutstandingLoanBalance(BigDecimal.valueOf(35000))
                     .transactionProcessingStrategyCode(FineractStyleLoanRepaymentScheduleTransactionProcessor.STRATEGY_CODE)
-                    .loanScheduleProcessingType(LoanScheduleProcessingType.HORIZONTAL.name()), 403);
+                    .loanScheduleProcessingType(LoanScheduleProcessingType.HORIZONTAL.name()), 400);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: It seems to me a business domain rule exception. I dont think this should be changed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 12]
```diff
@@ -1765,15 +1777,15 @@ private void validateTransactionProcessingStrategy(final String transactionProce
             if (LoanScheduleType.PROGRESSIVE.equals(loanProduct.getLoanProductRelatedDetail().getLoanScheduleType())) {
                 if (!LoanProductConstants.ADVANCED_PAYMENT_ALLOCATION_STRATEGY.equals(transactionProcessingStrategy)) {
                     // TODO: GeneralPlatformDomainRuleException vs PlatformApiDataValidationException
-                    throw new GeneralPlatformDomainRuleException(
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: This is a breaking change! Also I dont think it is correct! Please revert it

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java [hunk 13]
```diff
@@ -759,6 +759,14 @@ private void validateForCreate(final JsonElement element) {
                         graceOnInterestPayment, graceOnInterestCharged, recurringMoratoriumOnPrincipalPeriods, baseDataValidator);
             }
         });
+
+        validateOrThrow("loan", baseDataValidator -> {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationValidator.java]: This is  not needed hence the validator methods throws its own exceptions.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java [hunk 14]
```diff
@@ -172,6 +174,80 @@ public void uc2() {
...
+                    .transactionProcessingStrategyCode(LoanProductTestBuilder.ADVANCED_PAYMENT_ALLOCATION_STRATEGY)//
+                    .repaymentEvery(1)//
+                    .repaymentFrequencyType(2)//
+            ;//
+            callFailedRuntimeException = Assertions.assertThrows(CallFailedRuntimeException.class,
+                    () -> loanTransactionHelper.applyLoan(applicationRequest04));
+            assertEquals(400, callFailedRuntimeException.getResponse().code());
+            Assertions
+                    .assertTrue(callFailedRuntimeException.getMessage().contains("The parameter `expectedDisbursementDate` is mandatory."));
+        });
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanOriginationValidationTest.java:262]: @alberto-art3ch Please add test case that covers if loan product id is null, we got "The parameter `loanProductId` is mandatory."

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 15]
```diff
@@ -3784,8 +3784,7 @@ public void uc130() {
                             .loanScheduleProcessingType(LoanScheduleProcessingType.HORIZONTAL.name())));
 
             assertEquals(403, exception.getResponse().code());
-            assertTrue(exception.getMessage()
-                    .contains("error.msg.loan.repayment.strategy.can.not.be.different.than.advanced.payment.allocation"));
+            assertTrue(exception.getMessage().contains("error.msg.fixed.length.only.supported.for.advanced.payment.allocation"));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: This is incorrect! Fixed length validation has nothing to do here!

**Review by adamsaghy — DISMISSED** [review]
LGTM

**Comment by kjozsa** [comment]
@galovics do you have any concerns left with this PR?

**Review by adamsaghy — APPROVED** [review]
LGTM
